### PR TITLE
Lpfg 107 manage identities

### PR DESCRIPTION
**Pull request by smoga-git**

### CODE REVIEW

File: src/main/java/uk/gov/cshr/controller/IdentityController.java [hunk 1]
```diff
@@ -3,16 +3,119 @@
...
+            return "updateIdentity";
+        }
+
+        LOGGER.info("No identity found for uid {}", authenticationDetails.getCurrentUsername(), uid);
+        return "redirect:/management/identities";
+    }
+
+    @PostMapping("/identities/update")
+    public String identityUpdate(@RequestParam(value = "active", required = false) Boolean active, @RequestParam("roleId") ArrayList<String> roleId, @RequestParam("uid") String uid) {
+
```
Comment (m24murray) [inline src/main/java/uk/gov/cshr/controller/IdentityController.java]: On the 'Update Identity' screen, if the user doesn't select any roles and clicks update, they will receive a 400 error:

```
There was an unexpected error (type=Bad Request, status=400).
Required ArrayList parameter 'roleId' is not present
``` 

We could catch this exception, maybe by defining something like this in the controller:
```
@ExceptionHandler(MissingServletRequestParameterException.class)
    public String handleMissingParams(MissingServletRequestParameterException ex) {
        LOGGER.ERROR(ex.getParameterName() + " parameter is missing");
        return "redirect:/management/identities";
    }
```

Other solutions possible, so maybe lets have a talk

File: src/main/java/uk/gov/cshr/controller/IdentityController.java [hunk 2]
```diff
@@ -3,16 +3,119 @@
...
+    public String identities(Model model) {
+        LOGGER.info("Listing all identities");
+
+        Iterable<Identity> identities = identityService.findAll();
+        model.addAttribute("identities", identities);
+        return "identityList";
+    }
+
+    @GetMapping("/identities/update/{uid}")
+    public String identityEdit(Model model,
```
Comment (m24murray) [inline src/main/java/uk/gov/cshr/controller/IdentityController.java]: Maybe rename to identityUpdate for consistency?

Comment (smoga-git) [reply src/main/java/uk/gov/cshr/controller/IdentityController.java]: Yup

File: src/main/java/uk/gov/cshr/domain/Identity.java [hunk 3]
```diff
@@ -24,10 +24,10 @@
 
     private boolean active;
 
-    @ManyToMany(cascade = CascadeType.ALL, mappedBy = "identities", fetch = FetchType.EAGER)
+    @ManyToMany(cascade = CascadeType.ALL,  fetch = FetchType.EAGER)
```
Comment (m24murray) [inline src/main/java/uk/gov/cshr/domain/Identity.java]: Removing this mapping causes an error when deleting a role which has an identity attached

File: src/main/java/uk/gov/cshr/controller/IdentityController.java (hunk 1 above)
Comment (smoga-git) [reply src/main/java/uk/gov/cshr/controller/IdentityController.java]: I've just made it optional  and checked for it being set within the function loop

File: src/main/java/uk/gov/cshr/domain/Identity.java (hunk 3 above)
Comment (smoga-git) [reply src/main/java/uk/gov/cshr/domain/Identity.java]: But keeping it in means you can't update roles :D

File: src/main/java/uk/gov/cshr/controller/IdentityController.java [hunk 4]
```diff
@@ -1,18 +1,121 @@
...
+    private static final Logger LOGGER = LoggerFactory.getLogger(RoleController.class);
+
+    @Autowired
+    private IdentityService identityService;
+
+    @Autowired
+    private RoleService roleService;
+
+    @Autowired
+    private IdentityRepository identityRepository;
```
Comment (elxsy) [inline src/main/java/uk/gov/cshr/controller/IdentityController.java]: REDUNDANT - not being used

File: src/main/java/uk/gov/cshr/controller/IdentityController.java [hunk 5]
```diff
@@ -1,18 +1,121 @@
...
+    @PostMapping("/identities/update")
+    public String identityUpdate(@RequestParam(value = "active", required = false) Boolean active, @RequestParam(value = "roleId", required = false) ArrayList<String> roleId, @RequestParam("uid") String uid) {
+
+        // get identity to edit
+        Optional<Identity> optionalIdentity = identityService.getIdentity(uid);
+
+        if (optionalIdentity.isPresent()) {
+            Identity identity = optionalIdentity.get();
+
+            Set<Role> roleSet = new HashSet();
```
Comment (elxsy) [inline src/main/java/uk/gov/cshr/controller/IdentityController.java]: UNCHECKED - `new HashSet<>()`

File: src/main/java/uk/gov/cshr/controller/IdentityController.java [hunk 6]
```diff
@@ -1,18 +1,121 @@
...
+            // create roles from id
+            if (roleId != null) {
```
Comment (elxsy) [inline src/main/java/uk/gov/cshr/controller/IdentityController.java]: OPTIMIZE - Can we not use `roleId.forEach`?

Comment (smoga-git) [reply src/main/java/uk/gov/cshr/controller/IdentityController.java]: lambda expression is not superior than using enhanced for loop except with a parallel stream.

File: src/main/java/uk/gov/cshr/controller/IdentityController.java (hunk 5 above)
Comment (smoga-git) [reply src/main/java/uk/gov/cshr/controller/IdentityController.java]: Its a newly created HashSet ?

File: src/main/java/uk/gov/cshr/controller/IdentityController.java (hunk 4 above)
Comment (smoga-git) [reply src/main/java/uk/gov/cshr/controller/IdentityController.java]: done

File: src/main/java/uk/gov/cshr/controller/IdentityController.java (hunk 6 above)
Comment (elxsy) [reply src/main/java/uk/gov/cshr/controller/IdentityController.java]: Nothing to do with speed, readability and should prevent null checks too?

File: src/main/java/uk/gov/cshr/controller/IdentityController.java (hunk 5 above)
Comment (elxsy) [reply src/main/java/uk/gov/cshr/controller/IdentityController.java]: Static code checks does pick up on it

File: src/main/java/uk/gov/cshr/controller/IdentityController.java (hunk 6 above)
Comment (smoga-git) [reply src/main/java/uk/gov/cshr/controller/IdentityController.java]: Does not prevent null checks, exception thrown on nulls. As to readability, it's a pretty standard way of iteration?

**Pull request merged**